src/core/plugin/luapi_application.h

```diff
@@ -862,6 +862,7 @@ static int applib_addSplines(lua_State* L) {
     // -1 = splines
 
     size_t numSplines = lua_rawlen(L, -1);
+    ctrl->getDocument()->lock();
```

Comment:
I cannot opinion about the need of a lock or of the placement of them but I definitely I would like to see a RAII object for this.

In the `Document` class we could have a method:

```cpp
std::unique_lock<std::mutex> takeLock() { return std::unique_lock<std::mutex>(this->documentLock); }
```

Then, in the places where you wrote `...->getDocument()->lock();` you should have `auto lck = ...->getDocument()->takeLock();`
And with this all the `unlock` calls could be removed.

This is not only simpler and cleaner, but also safer. It is much harder to forget an unlock when the RAII object does it for you.

Reply:
Yea, fixing the locking issue globally definitely should be the goal in the end (I think this also is what they meant on gitter/matrix some time ago, not sure what exactly the progress is).
To avoid crashes in the meantime, I'm trying to add some locking to the luapi in its current state.